src/lib.rs

```diff
@@ -308,165 +173,288 @@ impl KafkaJsonToDelta {
-                        let should_complete_file = delta_writer.buffer_len()
-                            >= self.min_bytes_per_file
-                            || latency_timer.elapsed().as_secs() >= self.allowed_latency;
+                        delta_writer.write_record_batch(&record_batch).await?;
 
-                        debug!(
-                            "Should complete file? {}. delta_writer.buffer_len: {}. timer.elapsed: {}",
-                            should_complete_file,
-                            delta_writer.buffer_len(),
-                            latency_timer.elapsed().as_secs()
+                        info!(
+                            "Record batch written. Current latency is {} milliseconds",
+                            latency_timer.elapsed().as_millis()
                         );
 
-                        if should_complete_file {
-                            info!("Completing parquet file write.");
+                        // Finalize file and write Delta transaction
+                        if self.should_complete_file(&delta_writer, &latency_timer) {
+                            // Reset the latency timer to track allowed latency for the next file
+                            latency_timer = Instant::now();
 
-                            let add = delta_writer.complete_file(&partition_cols).await?;
-                            let action = deltalake::action::Action::add(add);
+                            info!("Writing parquet file.");
+
+                            // Lookup the last transaction version from the Delta Log
+                            let last_txn_version = delta_writer
+                                .last_transaction_version(self.app_id.as_str())
+                                .await;
+
+                            // Create the new WAL entry
+                            let wal_entry = if let Ok(Some(txn_version)) = last_txn_version {
+                                info!(
+                                    "Delta table last txn version for {} is {}.",
+                                    self.app_id, txn_version
+                                );
+                                let last_wal_entry =
+                                    wal.get_entry_by_transaction_id(txn_version).await?;
+                                last_wal_entry.prepare_next(&partition_offsets)
+                            } else {
+                                info!("Delta table does not contain a txn for app id {}. Starting write-ahead-log from 1.", self.app_id);
+                                WriteAheadLogEntry::new(
+                                    1,
+                                    TransactionState::Prepared,
+                                    None, /*delta table version starts out empty*/
+                                    partition_offsets.clone(),
+                                )
+                            };
+                            wal.put_entry(&wal_entry).await?;
```

Comment:
This `?` looks to me like another "if this errors the whole process dies" type thing, but I'm having a hard time parsing such a big change so I could be wrong :smile: